pybart/new_matcher.py

```diff
@@ -0,0 +1,330 @@
+# usage example:
+# conversions = [SomeConversion, SomeConversion1, ...]
+# matcher = Matcher(NamedConstraint(conversion.get_name(), conversion.get_constraint()) for conversion in conversions)
+#
+# def convert_sentence(sentence, matcher) -> sentence:
+#     while … # till convergance of the sentence
+#         m = matcher(doc)
+#         for conv_name in m.names():
+#             matches = m.matches_for(conv_name)
+#             transform(matches…) # TBD
+#         sentence = ...
+#
+# for doc in docs:
+#     convert_sentence(sentence, matcher)
+
+import itertools
+from collections import defaultdict
+from typing import NamedTuple, Sequence, Mapping, Any, List, Tuple
+import re
+import spacy
+from spacy.matcher import Matcher as SpacyMatcher
+from .constraints import *
+
+
+# function that checks that a sequence of Label constraints is satisfied
+# TODO - maybe move this function as a 'satisfied' method of the Label constraint? will simplify this code to
+#   "for constraint in label_constraints: if not label_constraint.satisfied(actual_labels): return False; return True"
```

Comment:
Sounds good.
And can simplify further to:
`return all((constraint.satisfied(actual_labels) for constraint in label_constraints))`

`

Reply:
@yoavg 
1. Actually this is not good for optimization (if we're interested), as we can break when one is not satisfied.
2. Now it's also a bit more complicated because I return a tuple of boolean + list of matched labels (and not - only boolean)

Reply:
1. re optimization, note that this uses `(x for x in ...)` and not `[x for x in ...]`, so its a generator, and `all` will also break early. so it's optimized. 
2. but this is indeed a problem.